modules/automode.cpp

```diff
@@ -0,0 +1,420 @@
+                   [=](const CString& sLine) { OnDelChansCommand(sLine); });
+        AddCommand("AddMasks", t_d("<user> <mask>,[mask] ..."),
+                   t_d("Adds masks to a user"),
+                   [=](const CString& sLine) { OnAddMasksCommand(sLine); });
+        AddCommand("DelMasks", t_d("<user> <mask>,[mask] ..."),
+                   t_d("Removes masks from a user"),
+                   [=](const CString& sLine) { OnDelMasksCommand(sLine); });
+        AddCommand("AddUser",
+                   t_d("<user> <hostmask>[,<hostmasks>...] <mode> [channels]"),
+                   t_d("Adds a user"),
+                   [=](const CString& sLine) { OnAddUserCommand(sLine); });
+        AddCommand("DelUser", t_d("<user>"), t_d("Removes a user"),
+                   [=](const CString& sLine) { OnDelUserCommand(sLine); });
+    }
+
+    bool OnLoad(const CString& sArgs, CString& sMessage) override {
+        // Load the users
+        for (MCString::iterator it = BeginNV(); it != EndNV(); ++it) {
+            const CString& sLine = it->second;
+            CAutoModeUser* pUser = new CAutoModeUser;
+            if (!pUser->FromString(sLine) ||
+                FindUser(pUser->GetUsername().AsLower())) {
+                delete pUser;
+            } else {
+                m_msUsers[pUser->GetUsername().AsLower()] = pUser;
+            }
+        }
+        return true;
+    }
+
+    ~CAutoModeMod() override {
+        for (const auto& it : m_msUsers) {
+            delete it.second;
+        }
+        m_msUsers.clear();
+    }
+
+    void OnJoin(const CNick& Nick, CChan& Channel) override {
+        // If we have permissions in this chan
+        if (Channel.HasPerm(CChan::Op) || Channel.HasPerm(CChan::HalfOp)) {
+            CheckAutoMode(Nick, Channel);
+        }
+    }
+
+    void OnMode(const CNick* pOpNick, const CNick& Nick, CChan& Channel,
+               bool bNoChange) {
+        if (Nick.GetNick() == GetNetwork()->GetIRCNick().GetNick()) {
+            const map<CString, CNick>& msNicks = Channel.GetNicks();
+            for (const auto& it : msNicks) {
+                if (!it.second.HasPerm(CChan::HalfOp)) {
+                    CheckAutoMode(it.second, Channel);
+                }
+            }
+        }
+    }
+
+    void OnModCommand(const CString& sLine) override {
+        CString sCommand = sLine.Token(0).AsUpper();
+        HandleCommand(sLine);
+    }
+
+    void OnAddUserCommand(const CString& sLine) {
+        CString sUser = sLine.Token(1);
+        CString sHost = sLine.Token(2);
+        CString sMode = sLine.Token(3);
+
+        if (sHost.empty()) {
+            PutModule(
+                t_s("Usage: AddUser <user> <hostmask>[,<hostmasks>...] <mode> "
+                    "[channels]"));
+        } else if (sMode != "q" && sMode != "a" && sMode != "o" &&
+                   sMode != "h" && sMode != "v") {
+            PutModule("<mode> must be one of the following: qaohv");
```

Comment:
You missed t_s

Reply:
It seems to work fine.. what is t_s anyway?